### ISSUE TITLE
SCOTCH + WW3 not running on orion 

### ISSUE THREAD

**Issue by JessicaMeixner-NOAA** [NOAA-EMC/WW3#885]
**Describe the bug**
PR #849 does not run on the NOAA Orion platform.   The crash details are described [here](https://github.com/NOAA-EMC/WW3/pull/849#issuecomment-1416154728).   

**To Reproduce**
Build and run on orion details are described [here](https://github.com/NOAA-EMC/WW3/pull/849#issuecomment-1416154728).   

**Expected behavior**
We should be able to run all regression tests on orion. 

**Additional context**

Current next steps are: 
* Continuing to follow up on running in debug mode
* Investigating using SCOTCH's findcmake files instead of adding them in WW3

**Comment by aerorahul** [comment]
I was able to build `scotch` on Orion via cmake:
The build script and installation of scotch on Orion is at:
`/work2/noaa/stmp/rmahajan/wavesWork/build_scotch.sh`

`scotch-v7.0.3` is built w/ MPI and installed at:
`/work2/noaa/stmp/rmahajan/wavesWork/install/scotch-v7.0.3-mpi`
All ctests for scotch pass.

Using this installation of scotch, I was able to build the [feature/scotchCI](https://github.com/JessicaMeixner-NOAA/WW3/tree/feature/scotchCI) branch at hash `7f0aa363` with the following modifications:
- remove `FindSCOTCH.cmake` and instead use `SCOTCHConfig.cmake` from the scotch installation above.
- update references to parmetis and adding targets to additional scotch library components
```
[feature/scotchCI ✚.2 …1] ❯❯❯ git diff model/
diff --git i/model/src/CMakeLists.txt w/model/src/CMakeLists.txt
index 7dd7d82a..9e8a9ef5 100644
--- i/model/src/CMakeLists.txt
+++ w/model/src/CMakeLists.txt
@@ -160,7 +160,11 @@ if("PDLIB" IN_LIST switches)
   if("SCOTCH" IN_LIST switches)
   find_package(SCOTCH REQUIRED)
   target_sources(ww3_lib PRIVATE ${pdlib_src})
-  target_link_libraries(ww3_lib PUBLIC PTSCOTCHparmetis::PTSCOTCHparmetis)
+  target_link_libraries(ww3_lib PUBLIC SCOTCH::ptscotchparmetisv3)
+  target_link_libraries(ww3_lib PUBLIC SCOTCH::scotcherr)
+  target_link_libraries(ww3_lib PUBLIC SCOTCH::scotcherrexit)
 elseif("METIS"  IN_LIST switches)
   find_package(ParMETIS REQUIRED)
   target_sources(ww3_lib PRIVATE ${pdlib_src})
```
The build script for WW3 is at: `/work2/noaa/stmp/rmahajan/wavesWork/build_ww3.sh`

**Comment by aliabdolali** [comment]
@aerorahul this is amazing.

**Comment by JessicaMeixner-NOAA** [comment]
@aerorahul is 
```
+  target_link_libraries(ww3_lib PUBLIC SCOTCH::ptscotchparmetisv3)
+  target_link_libraries(ww3_lib PUBLIC SCOTCH::ptscotchparmetisv3) 
```
 twice on purpose?

**Comment by aerorahul** [comment]
> @aerorahul is
> 
> ```
> +  target_link_libraries(ww3_lib PUBLIC SCOTCH::ptscotchparmetisv3)
> +  target_link_libraries(ww3_lib PUBLIC SCOTCH::ptscotchparmetisv3) 
> ```
> 
> twice on purpose?

sorry `copy-paste` error.
I updated the branch and rebuilt.  All good.  The above comment has been updated.

**Comment by arunchawla-NOAA** [comment]
Thank you Rahul! 

Waves team does this take care of your build issues? Will this work for HERA too?

**Comment by JessicaMeixner-NOAA** [comment]
@arunchawla-NOAA  still TBD if it solves the failure in the runs or not on orion.

**Comment by JessicaMeixner-NOAA** [comment]
The ci test failed: 

<img width="1012" alt="Screen Shot 2023-02-10 at 12 33 36 PM" src="https://user-images.githubusercontent.com/22549231/218158143-9f9244e9-966f-4d82-9817-fb7edbf5d192.png">

perhaps is a scotch version issue? But I'll dig into this after I dig into lunch :D 

The updates from @aerorahul are now committed to my branch here: https://github.com/JessicaMeixner-NOAA/WW3/tree/feature/usescotchcmake

**Comment by MatthewMasarik-NOAA** [comment]
> Will this work for HERA too?

@arunchawla-NOAA this is also TBD.  Currently I can run the regtests and the GFSv17 candidate grid canned case on hera without error, except the known scaling issues.  It's unknown right now if this fix is related to the scaling issue on hera.

**Comment by aliabdolali** [comment]
@aerorahul   
looking at /work2/noaa/stmp/rmahajan/wavesWork/build_ww3.sh, the following does the trick
```
module use  ./modulefiles
 module load gcc/10.2.0
```
But for Hera, we do not have it and I could simply load gnu and it worked. Do we need to extend build_ww3.sh to check for gnu for hera and gcc for orion?

**Comment by arunchawla-NOAA** [comment]
This may be a silly question and probably not one to ask on a late Friday evening but are you able to build outside the CI test and run the model with the SCOTCH library?

**Comment by aliabdolali** [comment]
@arunchawla-NOAA on Hera yes, we can run ww3 build with both cmake and gnu make.
On Orion, no until Rahul's fix. My solution to fix gcc was a fix to compile scotch. With Rahul's trick, we can load gcc without manipulation. Now I bet ww3 runs with scotch (we are not talking about beyond 2k cpus).
We will investigate further next week

**Comment by aliabdolali** [comment]
I tried Rahul's way of compiling SCOTCH on orion, it was successful, and WW3 was compiled as well, but the run failed.
Note that my way of SCOTCH build ended up with the same error.

**Comment by MatthewMasarik-NOAA** [comment]
Update:  Working with @aerorahul this morning the build issue on Orion with SCOTCH has been resolved.  This was by following up the 1) lead I posted on pthreads last friday.  Work on the scaling for both orion and hera are underway.  It sounds like this clue has also been a break-through for @aliabdolali and co.  It will be nice to have independent testing to verify our solution.

**Comment by aliabdolali** [comment]
Awesome Matt.
We were able to run the case with 8000 cpu on ERDC machine.
Please provide how to compile scotch properly.

**Comment by aerorahul** [comment]
@aliabdolali 
configure `scotch` with `-DTHREADS=OFF` and build.

**Comment by aliabdolali** [comment]
@aerorahul do you have a build script? I wanna document them and we also need to inform the same requirements for WCOSS2 (after our final checks).

**Comment by aerorahul** [comment]
Copy the modified GCC module from: 
```bash
>> cp -rv /work2/noaa/stmp/rmahajan/wavesWork/modulefiles .
```

Save the following as `build_scotch.sh` and give it `exe` permissions `chmod +x build_scotch.sh`

```bash
#!/bin/bash

set -eux

BASEDIR="${PWD}"

name="scotch"
version=${VERSION:-"v7.0.3"}
INSTALL_PREFIX=${INSTALL_PREFIX:-${BASEDIR}/install/$name-$version}

[[ -d ${BASEDIR}/$name-$version ]] && rm -rf ${BASEDIR}/$name-$version
[[ ! -d ${BASEDIR}/$name-$version ]] && git clone -b $version https://gitlab.inria.fr/scotch/scotch ${BASEDIR}/$name-$version
[[ ! -d ${BASEDIR}/$name-$version ]] && ( echo "${BASEDIR}/$name-$version does not exist, ABORT!"; exit 1)

set +x
module load cmake/3.22.1
module use  /work/noaa/epic-ps/hpc-stack/libs/intel/2022.1.2/modulefiles/stack
module load hpc/1.2.0
module load hpc-intel/2022.1.2
module load hpc-impi/2022.1.2
if [[ ${USE_GCC:-"YES"} = "YES" ]]; then
  # SCOTCH uses newer GCC headers that are not available unless this is loaded
  module use  ${BASEDIR}/modulefiles
  module load gcc/10.2.0
fi
module list
set -x

#GXX_ROOT=/apps/gcc-10.2.0/gcc-10.2.0
export CC=mpiicc     # icc makes no difference as hpc-impi is loaded
export FC=mpiifort   # ifort makes no difference as hpc-impi is loaded

[[ -d $INSTALL_PREFIX ]] && rm -rf $INSTALL_PREFIX

BUILD_DIR=${BUILD_DIR:-"build/$name-$version"}
[[ ${BUILD_CLEAN:-"YES"} = "YES" ]] && rm -rf ${BUILD_DIR}
mkdir -p ${BUILD_DIR} && cd ${BUILD_DIR}

cmake ${BASEDIR}/$name-$version \
      ${CMAKE_OPTIONS:-} \
      -DCMAKE_INSTALL_PREFIX=${INSTALL_PREFIX} \
      -DCMAKE_BUILD_TYPE=${BUILD_TYPE:-"Release"}
make -j ${BUILD_JOBS:-1} VERBOSE=${BUILD_VERBOSE:-}
[[ ${CTEST_RUN:-"NO"} = "YES" ]] && ctest -j ${CTEST_JOBS:-1}
make install
```

```bash
>> CMAKE_OPTIONS="-DTHREADS=OFF" ./build_scotch.sh |& tee log.scotch
```

`scotch` will be installed under `./install/scotch-v7.0.3`

**Comment by arunchawla-NOAA** [comment]
where are we with this? We need to make a request for how the SCOTCH library will be installed in operations so need to know where we stand with this

**Comment by aliabdolali** [comment]
We need to see if the build proposed by Rahul and Matt passes the 2400 cores on RDHPC.

**Comment by aerorahul** [comment]
@aliabdolali 
IMO, there are 2 issues:
1. usability of scotch (on Orion) within WW3
2. scalability of ww3 (possibly w/ or w/o scotch)

Can we address this issue titled "SCOTCH + WW3 not running on Orion".  Has that been adequately addressed?

**Comment by aliabdolali** [comment]
> @aliabdolali IMO, there are 2 issues:
> 
> 1. usability of scotch (on Orion) within WW3
> 2. scalability of ww3 (possibly w/ or w/o scotch)
> 
> Can we address this issue titled "SCOTCH + WW3 not running on Orion". Has that been adequately addressed?

My understanding is that WW3 + SCOTCH with the build you and Matt did is stable for our tiny regtests on 24 cpus. 
The problem is with >2400 cpus on both Hera and Orion and possibly WCOSS2.

Regarding the scalability of ww3 w/o scotch, the model is scalable, isn't it?

**Comment by MatthewMasarik-NOAA** [comment]
Hi all,

1. has been addressed.
2.  is ongoing.

Note that 2. is not holding the SCOTCH PR back.  I've made a PR with the new SCOTCH builds for both hera and orion, it is a draft until the regtests come back.  Orion passed and output will be added shortly.  Hera I am working on. 

We had the repo manager meeting this morning.  Several mgrs will be testing, and that have until Mar 1.  Regtest output on RDHPCS machines will be added today or tomorrow at the latest. 

@arunchawla-NOAA 
For how it will be built in production, I will address that by building and running on Acorn.  I just got my account there recently and will be working on that.

**bottomline** -  SCOTCH PR is in need of community approval which we will close Mar 1st, then merge.  For the scaling, I'm continuing to look into this, though the SCOTCH PR is _not_ dependent on this aspect.
